File: src/installer.py
"""
ADB Tools Auto-Installer Module
Downloads and sets up Android Platform Tools on first run
"""

import os
import sys
import platform
import zipfile
import tarfile
import shutil
import stat
import subprocess
from pathlib import Path
from typing import Optional, Callable

try:
    import requests
    from tqdm import tqdm
except ImportError:
    requests = None
    tqdm = None


class ADBInstaller:
    """Handles automatic download and installation of ADB tools."""

    # Official Google Platform Tools download URLs
    DOWNLOAD_URLS = {
        "Windows": "https://dl.google.com/android/repository/platform-tools-latest-windows.zip",
        "Linux": "https://dl.google.com/android/repository/platform-tools-latest-linux.zip",
        "Darwin": "https://dl.google.com/android/repository/platform-tools-latest-darwin.zip",
    }

    def __init__(self, install_dir: Optional[Path] = None):
        """
        Initialize the installer.

        Args:
            install_dir: Directory to install ADB tools. Defaults to ./android-tools/
        """
        self.system = platform.system()
        self.install_dir = install_dir or Path(__file__).parent.parent / "android-tools"
        self.platform_tools_dir = self.install_dir / "platform-tools"
        self.adb_path = self._get_adb_executable_path()

    def _get_adb_executable_path(self) -> Path:
        """Get the path to the ADB executable based on OS."""
        if self.system == "Windows":
            return self.platform_tools_dir / "adb.exe"
        return self.platform_tools_dir / "adb"

    def is_installed(self) -> bool:
        """Check if ADB is already installed in our directory."""
        return self.adb_path.exists()

    def is_system_adb_available(self) -> bool:
        """Check if ADB is available in system PATH."""
        try:
            result = subprocess.run(
                ["adb", "version"],
                capture_output=True,
                text=True,
                timeout=5
            )
            return result.returncode == 0
        except (subprocess.SubprocessError, FileNotFoundError):
            return False

    def get_adb_command(self) -> str:
        """Get the ADB command to use (local or system)."""
        if self.is_installed():
            return str(self.adb_path)
        elif self.is_system_adb_available():
            return "adb"
        return str(self.adb_path)  # Will need to install first

    def download_with_progress(
        self,
        url: str,
        dest: Path,
        progress_callback: Optional[Callable[[int, int], None]] = None
    ) -> bool:
        """
        Download a file with progress tracking.

        Args:
            url: URL to download from
            dest: Destination file path
            progress_callback: Optional callback(downloaded, total) for progress updates

        Returns:
            True if download successful, False otherwise
        """
        if requests is None:
            print("Error: 'requests' library not installed. Run: pip install requests")
            return False

        try:
            response = requests.get(url, stream=True, timeout=30)
            response.raise_for_status()

            total_size = int(response.headers.get('content-length', 0))
            downloaded = 0

            dest.parent.mkdir(parents=True, exist_ok=True)

            with open(dest, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    if chunk:
                        f.write(chunk)
                        downloaded += len(chunk)
                        if progress_callback:
                            progress_callback(downloaded, total_size)

            return True

        except requests.RequestException as e:
            print(f"Download error: {e}")
            return False

    def extract_archive(self, archive_path: Path, extract_to: Path) -> bool:
        """
        Extract a zip or tar archive.

        Args:
            archive_path: Path to the archive file
            extract_to: Directory to extract to

        Returns:
            True if extraction successful, False otherwise
        """
        try:
            extract_to.mkdir(parents=True, exist_ok=True)

            if archive_path.suffix == '.zip':
                with zipfile.ZipFile(archive_path, 'r') as zf:
                    zf.extractall(extract_to)
            elif archive_path.suffix in ['.tar', '.gz', '.tgz']:
                with tarfile.open(archive_path, 'r:*') as tf:
                    tf.extractall(extract_to)
            else:
                print(f"Unknown archive format: {archive_path.suffix}")
                return False

            return True

        except (zipfile.BadZipFile, tarfile.TarError) as e:
            print(f"Extraction error: {e}")
            return False

    def set_executable_permissions(self) -> None:
        """Set executable permissions on ADB binaries (Linux/macOS)."""
        if self.system == "Windows":
            return

        executables = ["adb", "fastboot", "mke2fs", "e2fsdroid"]
        for exe in executables:
            exe_path = self.platform_tools_dir / exe
            if exe_path.exists():
                current_mode = exe_path.stat().st_mode
                exe_path.chmod(current_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)

    def install(
        self,
        progress_callback: Optional[Callable[[str, int, int], None]] = None
    ) -> bool:
        """
        Download and install ADB tools.

        Args:
            progress_callback: Optional callback(status, downloaded, total)

        Returns:
            True if installation successful, False otherwise
        """
        if self.is_installed():
            return True

        if self.system not in self.DOWNLOAD_URLS:
            print(f"Unsupported operating system: {self.system}")
            return False

        url = self.DOWNLOAD_URLS[self.system]
        archive_name = f"platform-tools-{self.system.lower()}.zip"
        archive_path = self.install_dir / archive_name

        # Create install directory
        self.install_dir.mkdir(parents=True, exist_ok=True)

        # Download
        def download_progress(downloaded: int, total: int):
            if progress_callback:
                progress_callback("Downloading Android Platform Tools...", downloaded, total)

        if not self.download_with_progress(url, archive_path, download_progress):
            return False

        # Extract
        if progress_callback:
            progress_callback("Extracting...", 0, 0)

        if not self.extract_archive(archive_path, self.install_dir):
            return False

        # Set permissions
        self.set_executable_permissions()

        # Clean up archive
        try:
            archive_path.unlink()
        except OSError:
            pass

        # Verify installation
        if not self.is_installed():
            print("Installation failed: ADB executable not found after extraction")
            return False

        if progress_callback:
            progress_callback("Installation complete!", 100, 100)

        return True

    def uninstall(self) -> bool:
        """Remove installed ADB tools."""
        try:
            if self.install_dir.exists():
                shutil.rmtree(self.install_dir)
            return True
        except OSError as e:
            print(f"Uninstall error: {e}")
            return False

    def get_version(self) -> Optional[str]:
        """Get the installed ADB version."""
        adb_cmd = self.get_adb_command()
        try:
            result = subprocess.run(
                [adb_cmd, "version"],
                capture_output=True,
                text=True,
                timeout=5
            )
            if result.returncode == 0:
                # Parse version from output
                for line in result.stdout.split('\n'):
                    if 'Android Debug Bridge version' in line:
                        return line.split('version')[-1].strip()
            return None
        except (subprocess.SubprocessError, FileNotFoundError):
            return None


def ensure_adb_installed(install_dir: Optional[Path] = None) -> str:
    """
    Convenience function to ensure ADB is installed and return the command path.

    Args:
        install_dir: Optional installation directory

    Returns:
        Path to ADB executable

    Raises:
        RuntimeError: If ADB cannot be installed
    """
    installer = ADBInstaller(install_dir)

    if installer.is_installed() or installer.is_system_adb_available():
        return installer.get_adb_command()

    print("ADB not found. Downloading Android Platform Tools...")

    def progress(status: str, downloaded: int, total: int):
        if total > 0:
            pct = (downloaded / total) * 100
            print(f"\r{status} {pct:.1f}%", end="", flush=True)
        else:
            print(f"\r{status}", end="", flush=True)

    if installer.install(progress):
        print("\nADB installed successfully!")
        return installer.get_adb_command()
    else:
        raise RuntimeError("Failed to install ADB tools")


if __name__ == "__main__":
    # Test installation
    try:
        adb_path = ensure_adb_installed()
        print(f"ADB available at: {adb_path}")

        installer = ADBInstaller()
        version = installer.get_version()
        if version:
            print(f"ADB version: {version}")
    except RuntimeError as e:
        print(f"Error: {e}")
        sys.exit(1)

File: src/installer.rs
//! ADB Platform Tools installer - auto-downloads from Google

use anyhow::{anyhow, Result};
use std::fs::{self, File};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};

/// Download URLs for platform tools
const PLATFORM_TOOLS_WINDOWS: &str = "https://dl.google.com/android/repository/platform-tools-latest-windows.zip";
const PLATFORM_TOOLS_LINUX: &str = "https://dl.google.com/android/repository/platform-tools-latest-linux.zip";
const PLATFORM_TOOLS_MACOS: &str = "https://dl.google.com/android/repository/platform-tools-latest-darwin.zip";

/// Get the data directory for the application
pub fn get_data_dir() -> PathBuf {
    dirs::data_local_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("android-supertool")
}

/// Get the path where ADB should be installed
pub fn get_adb_dir() -> PathBuf {
    get_data_dir().join("platform-tools")
}

/// Get the path to the ADB executable
pub fn get_adb_path() -> PathBuf {
    let adb_dir = get_adb_dir();
    if cfg!(windows) {
        adb_dir.join("adb.exe")
    } else {
        adb_dir.join("adb")
    }
}

/// Check if ADB is already installed
pub fn is_adb_installed() -> bool {
    get_adb_path().exists()
}

/// Check if ADB is available in system PATH
pub fn is_adb_in_path() -> Option<PathBuf> {
    which::which("adb").ok()
}

/// Get the best available ADB path
pub fn get_best_adb_path() -> Option<PathBuf> {
    // First check our local installation
    if is_adb_installed() {
        return Some(get_adb_path());
    }

    // Then check system PATH
    is_adb_in_path()
}

/// Installer for ADB Platform Tools
pub struct Installer {
    progress_callback: Option<Box<dyn Fn(f32, &str) + Send>>,
}

impl Installer {
    pub fn new() -> Self {
        Self {
            progress_callback: None,
        }
    }

    /// Set a callback for progress updates
    pub fn with_progress<F>(mut self, callback: F) -> Self
    where
        F: Fn(f32, &str) + Send + 'static,
    {
        self.progress_callback = Some(Box::new(callback));
        self
    }

    fn report_progress(&self, progress: f32, message: &str) {
        if let Some(ref callback) = self.progress_callback {
            callback(progress, message);
        }
    }

    /// Download and install ADB Platform Tools
    pub fn install(&self) -> Result<PathBuf> {
        // Create data directory
        let data_dir = get_data_dir();
        fs::create_dir_all(&data_dir)?;

        self.report_progress(0.0, "Starting download...");

        // Get download URL for current OS
        let url = if cfg!(target_os = "windows") {
            PLATFORM_TOOLS_WINDOWS
        } else if cfg!(target_os = "macos") {
            PLATFORM_TOOLS_MACOS
        } else {
            PLATFORM_TOOLS_LINUX
        };

        // Download the zip file
        let zip_path = data_dir.join("platform-tools.zip");
        self.download_file(url, &zip_path)?;

        self.report_progress(0.7, "Extracting...");

        // Extract the zip
        self.extract_zip(&zip_path, &data_dir)?;

        // Clean up zip file
        let _ = fs::remove_file(&zip_path);

        // Make ADB executable on Unix
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let adb_path = get_adb_path();
            if adb_path.exists() {
                let mut perms = fs::metadata(&adb_path)?.permissions();
                perms.set_mode(0o755);
                fs::set_permissions(&adb_path, perms)?;
            }

            // Also make fastboot executable
            let fastboot_path = get_adb_dir().join("fastboot");
            if fastboot_path.exists() {
                let mut perms = fs::metadata(&fastboot_path)?.permissions();
                perms.set_mode(0o755);
                fs::set_permissions(&fastboot_path, perms)?;
            }
        }

        self.report_progress(1.0, "Installation complete!");

        let adb_path = get_adb_path();
        if adb_path.exists() {
            Ok(adb_path)
        } else {
            Err(anyhow!("ADB installation failed - executable not found"))
        }
    }

    /// Download a file with progress reporting
    fn download_file(&self, url: &str, dest: &Path) -> Result<()> {
        self.report_progress(0.1, "Connecting to Google servers...");

        let response = reqwest::blocking::get(url)?;

        if !response.status().is_success() {
            return Err(anyhow!("Download failed: HTTP {}", response.status()));
        }

        let total_size = response.content_length().unwrap_or(0);
        let mut downloaded: u64 = 0;

        let mut file = File::create(dest)?;
        let mut reader = response;

        let mut buffer = [0u8; 8192];
        loop {
            let bytes_read = reader.read(&mut buffer)?;
            if bytes_read == 0 {
                break;
            }

            file.write_all(&buffer[..bytes_read])?;
            downloaded += bytes_read as u64;

            if total_size > 0 {
                let progress = 0.1 + (downloaded as f32 / total_size as f32) * 0.6;
                let mb_downloaded = downloaded as f32 / 1_048_576.0;
                let mb_total = total_size as f32 / 1_048_576.0;
                self.report_progress(
                    progress,
                    &format!("Downloading... {:.1}/{:.1} MB", mb_downloaded, mb_total),
                );
            }
        }

        Ok(())
    }

    /// Extract a zip file
    fn extract_zip(&self, zip_path: &Path, dest_dir: &Path) -> Result<()> {
        let file = File::open(zip_path)?;
        let mut archive = zip::ZipArchive::new(file)?;

        let total_files = archive.len();

        for i in 0..archive.len() {
            let mut file = archive.by_index(i)?;
            let outpath = match file.enclosed_name() {
                Some(path) => dest_dir.join(path),
                None => continue,
            };

            if file.name().ends_with('/') {
                fs::create_dir_all(&outpath)?;
            } else {
                if let Some(parent) = outpath.parent() {
                    if !parent.exists() {
                        fs::create_dir_all(parent)?;
                    }
                }
                let mut outfile = File::create(&outpath)?;
                std::io::copy(&mut file, &mut outfile)?;
            }

            let progress = 0.7 + (i as f32 / total_files as f32) * 0.25;
            self.report_progress(progress, &format!("Extracting files... {}/{}", i + 1, total_files));
        }

        Ok(())
    }
}

impl Default for Installer {
    fn default() -> Self {
        Self::new()
    }
}
